Draft rate limiter with example usage

camgunz opened the pull request:
Martin's comments have made me realize I should say this is just to elicit feedback; we'd at least add metrics and tests before actually merging this, and porting Postman's rate limiting would be done separately.

<!--
Please create an issue to collect feedback prior to feature additions. Please also reference that issue in any PRs.
If possible try to keep PRs scoped to one feature, and add tests for new features.
-->

### Description:
Adds a base rate limiter implementation and a usage example for demonstration.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

camgunz commented on pkg/common/rate_limiter.go:
```diff
@@ -0,0 +1,122 @@
+		//		    limit sleeps for a duration instead of until a specific time, but
+		//			  I haven't thought through that.
+		if err := lim.Execute(ctx, req, now); err != nil {
+			return nil, fmt.Errorf(
```
Oh definitely 💯 . One of the things I'm hoping this gives us is unified metrics, etc. on rate limiting. I think we wouldn't merge something like this w/o that, at least I wouldn't want to.

---

Something I was thinking about is how useful it is for limits to have names. Like, I had that thought and I was waiting for validation that it was correct, and I think this is it. The options I can think of for this are:

- instead of accepting (effectively) an array of limits in `NewRateLimiter`, use a `map[string]RateLimit` parameter instead
- require `RateLimit` to have a `.Name` method

If it's possible to have the same `RateLimit` in multiple `RateLimiter`s, I don't really know what's useful here. It'd be annoying if two `RateLimit`s had the same name (`.Name` would allow this unless we explicitly check for it), but maybe it's also annoying for the same `RateLimit` to have different names in different `RateLimiter`s.

OK thinking through it, I'm leaning towards `map[string]RateLimit`:
- it's easy to trace what name is assigned to a `RateLimit` in calls to `NewRateLimiter`
- naming uniqueness is strongly signaled and subsequently enforced
- names may be better, i.e. you're more likely to get "max 500r/month" as a name than "TokenBucketRateLimiter"

Yeah actually that last point pushes me over the edge, OK

camgunz commented:
> I feel quite strongly that anything that ends up anywhere with a name like common should be well covered with unit tests and (hopefully) examples for how clients of this library would write integration tests using this library. Is there anything blocking us from doing that here?

Oh this would be lousy with tests for sure. I'm currently wondering if it should be its own package so we can put common `RateLimit`s here and test them thoroughly as well.

rosecodym commented on pkg/common/rate_limiter.go:
```diff
@@ -0,0 +1,122 @@
+	wg.Wait()
+
+	if updateError != nil {
+		return nil, fmt.Errorf("error updating rate limits: %w", updateError)
```
What kinds of errors might we see here?

camgunz replied:
Interesting question... hmmm, off the top of my head:
- unexpected HTTP status
- looked for a header/headers but didn't get one/them
- found a header(s) but value was bad
  - this could be anything from "'90 seconds' is not an integer" to "the API asked me to wait six years... I'm not doing that"
- updating the state (e.g. if we're persisting it somewhere) failed
- context timed out

camgunz commented on pkg/common/rate_limiter.go:
```diff
@@ -0,0 +1,122 @@
+// thus can be used in more than one rate limiter. For example, if an API has 2
+// endpoints, one accepts 5r/s and another accepts 1r/s, but both have a limit
+// of total 500r/month, the policy implementing the 500r/month limit should be
+// able to be used in both of the 2 rate limiters for the 5r/s and 1r/s limits.
```
Haha ok let me :hammer: on it a little

camgunz commented on pkg/common/rate_limiter.go:
```diff
@@ -0,0 +1,122 @@
+// - Create a RateLimiter with its limits
+// - Call .Do instead of what you would normally call to make a request
+// - Process the response (returned from .Do) as normal
+type RateLimiter struct {
```
(cc: @rosecodym) I started to `%s/RateLimit/SourceIntegrationRateLimit/g` but then I thought what if instead of `common` we put this in `sources`?

camgunz commented on pkg/common/rate_limiter.go:
```diff
@@ -0,0 +1,122 @@
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+)
+
+// RateLimit represents a rate limiting implementation. A rate limiter
+// comprises 0 or more limits. Policies should be goroutine safe.
```
Oh no this is perfect (also I got rid of "policy" and that was an imperfect editing step). Let me tighten it up a bit 👍🏻

camgunz commented:
> And why does a limit being used for a single API mean that it can be used in more than one rate limiter? (Why would a limit used for multiple APIs not be usable in more than one rate limiter?)

I renamed some stuff and changed some comments that hopefully make this clearer. But, yeah overall I'm struggling to demonstrate how I want someone to use/implement this. The Postman example here (i.e. ratelimiter-per-source) is what I'm thinking, but someone could do a centralized one a la:

```go
// in some far away file
var CloudRateLimiter *RateLimiter

type CloudRateLimits struct {
// ...
}

func (crl *CloudRateLimits) MaybeWait(
  ctx *context.Context,
  req *http.Request,
  makeRequest func() (*http.Request, error),
  now time.Time,
) (*http.Request, error) {
  switch req.URL.Host {
    case "api.postman.com":
      return crl.MaybeWaitPostman(ctx, req, makeRequest, now)
    case api.jira.com":
     return crl.MaybeWaitJira(ctx, req, makeRequest, now)
    // ...
  }
}  

func init() {
  CloudRateLimiter = NewRateLimiter(NewCloudRateLimits())
}
```

Maybe someone would argue the centralized version is better? I think these are the downsides:
- Modifying a rate limit for a Source is a lot more likely to break rate limiting for another/all other Sources
- Locality is by definition a lot worse
- An errant RateLimit for one Source is a lot more likely to break rate limiting for another/all Sources
- It entwines three separate things, making most of the code in the module/file irrelevant to you (except it is relevant, because any `RateLimit` can break your rate limiting):
  - Common rate limiting functionality (the loop in `RateLimiter`)
  - The limits for the API/Source you're interested in
  - The limits for all the other APIs/Sources you're not interested in

One way to more or less prohibit (at least strongly signal against) this is to accept a `hostname` parameter when building `RateLimiter` and `RateLimit`. Its purpose wouldn't be functional, rather it'd be strictly for safety: we'd simply check that all the `RateLimit`s have the same hostname, and that all requests sent through `.Do` have the same hostname. People might feel like it's bowling with bumpers, but the security term is belt and suspenders, which I'm pro.

Let me play around w/ that and work on the naming a little more.

camgunz commented on pkg/common/rate_limiter.go:
```diff
@@ -0,0 +1,122 @@
+// of total 500r/month, the policy implementing the 500r/month limit should be
+// able to be used in both of the 2 rate limiters for the 5r/s and 1r/s limits.
+type RateLimit interface {
+	// Execute and update execute and update a policy, respectively. These
```
Good idea yeah; gave this a shot 👍🏻

camgunz commented:
OK I did a bunch of stuff; happy to undo it or do different things. This is fun!

camgunz commented:
Mothballing for now 🦋
